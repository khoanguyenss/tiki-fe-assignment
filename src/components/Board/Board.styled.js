import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const Wrapper = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const HeaderWrapper = styled.div`
  margin-bottom: 25px;
`;

export const RestartBtn = styled.button`
  font-size: 18px;
  margin-left: 15px;
  background: unset;
  border: unset;
  outline: unset;
  
  &:hover {
    text-decoration: underline;
    cursor:pointer;
  }
`;

export const HomeLink = styled(Link)`
  font-size: 18px;
  color: #000;
  text-decoration: none;
  
  &:hover {
    text-decoration: underline;
  }
`;

export const BoardWrapper = styled.div`
`;

export const ErrorMsg = styled.div`
  color: crimson;
  font-size: 24px;
`;
