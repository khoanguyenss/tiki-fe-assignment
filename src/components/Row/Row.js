import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { RowWrapper } from './Row.styled';
import Box from '../../containers/Box';

const Row = memo(({ row }) => (
  <RowWrapper>
    {row.map((box, index) => <Box box={box} key={index} />)}
  </RowWrapper>
));

Row.propTypes = {
  row: PropTypes.arrayOf(
    PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number,
      data: PropTypes.number,
      isOpen: PropTypes.bool,
    }),
  ).isRequired,
};

export default Row;
