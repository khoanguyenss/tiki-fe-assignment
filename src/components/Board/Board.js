import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { BoardWrapper } from './Board.styled';
import Row from '../Row';

const Board = ({ fetchBoard, board, isLoading }) => {
  useEffect(() => {
    fetchBoard();

    return undefined; // no un-sub event
  }, []);

  if (isLoading) {
    return (
      <div>
        Preparing game for you
      </div>
    );
  }

  return (
    <BoardWrapper>
      {board.map((row, index) => (<Row row={row} key={index} />))}
    </BoardWrapper>
  );
};

Board.propTypes = {
  fetchBoard: PropTypes.func.isRequired,
  board: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
        data: PropTypes.number,
        isOpen: PropTypes.bool,
      }),
    ),
  ).isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default Board;
