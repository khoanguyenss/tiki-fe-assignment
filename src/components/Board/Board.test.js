import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import Board from './Board';
import { RestartBtn } from './Board.styled';
import Row from '../Row';

describe('<Board />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const fetchBoardMock = jest.fn();

  const Component = mount(
    <Router>
      <Board
        fetchBoard={fetchBoardMock}
        board={[[], [], [], []]}
        isLoading={false}
        errorMsg={null}
      />
    </Router>
    ,
  );

  it('should render', () => {
    expect(Component.length).toBe(1);
  });
  it('match snapshot', () => {
    expect(Component).toMatchSnapshot();
  });
  it('should render 4 rows', () => {
    const RestartBtnComp = Component.find(RestartBtn).at(0);
    RestartBtnComp.simulate('click');
    expect(fetchBoardMock).toBeCalledTimes(1);
  });
  it('should call fetchBoardMock once', () => {
    // trigger useEffect
    act(() => {
      mount(
        <Router>
          <Board
            fetchBoard={fetchBoardMock}
            board={[]}
            isLoading={false}
            errorMsg={null}
          />
        </Router>
        ,
      );
    });
    expect(fetchBoardMock).toBeCalledTimes(1);
  });
  it('should trigger fetchBoardMock when Restart button is clicked', () => {
    const RowList = Component.find(Row);
    expect(RowList.length).toBe(4);
  });
});
