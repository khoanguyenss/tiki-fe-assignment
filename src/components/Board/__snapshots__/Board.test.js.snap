// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> match snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Board
      board={
        Array [
          Array [],
          Array [],
          Array [],
          Array [],
        ]
      }
      errorMsg={null}
      fetchBoard={[MockFunction]}
      isLoading={false}
    >
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "bEeQEn",
                "rules": Array [
                  "
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bdVaJa bEeQEn"
          >
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": false,
                      "lastClassName": "ixfxJY",
                      "rules": Array [
                        "
  margin-bottom: 25px;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-bwzfXH ixfxJY"
                >
                  <Styled(Link)
                    to="/"
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": false,
                            "lastClassName": "iVQJly",
                            "rules": Array [
                              "
  font-size: 18px;
  color: #000;
  text-decoration: none;
  
  &:hover {
    text-decoration: underline;
  }
",
                            ],
                          },
                          "displayName": "Styled(Link)",
                          "foldedComponentIds": Array [],
                          "propTypes": Object {
                            "innerRef": [Function],
                            "onClick": [Function],
                            "replace": [Function],
                            "target": [Function],
                            "to": [Function],
                          },
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "displayName": "Link",
                            "propTypes": Object {
                              "innerRef": [Function],
                              "onClick": [Function],
                              "replace": [Function],
                              "target": [Function],
                              "to": [Function],
                            },
                            "render": [Function],
                          },
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      to="/"
                    >
                      <Link
                        className="sc-bxivhb iVQJly"
                        to="/"
                      >
                        <LinkAnchor
                          className="sc-bxivhb iVQJly"
                          href="/"
                          navigate={[Function]}
                        >
                          <a
                            className="sc-bxivhb iVQJly"
                            href="/"
                            onClick={[Function]}
                          >
                            Home
                          </a>
                        </LinkAnchor>
                      </Link>
                    </StyledComponent>
                  </Styled(Link)>
                  <styled.button
                    onClick={[Function]}
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-htpNat",
                            "isStatic": false,
                            "lastClassName": "gACCbB",
                            "rules": Array [
                              "
  font-size: 18px;
  margin-left: 15px;
  background: unset;
  border: unset;
  outline: unset;
  
  &:hover {
    text-decoration: underline;
    cursor:pointer;
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-htpNat",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                    >
                      <button
                        className="sc-htpNat gACCbB"
                        onClick={[Function]}
                      >
                        Restart
                      </button>
                    </StyledComponent>
                  </styled.button>
                </div>
              </StyledComponent>
            </styled.div>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-ifAKCX",
                      "isStatic": false,
                      "lastClassName": "bmKgXx",
                      "rules": Array [
                        "
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-ifAKCX",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-ifAKCX bmKgXx"
                >
                  <Memo()
                    key="0"
                    row={Array []}
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bZQynM",
                              "isStatic": false,
                              "lastClassName": "kzlsZH",
                              "rules": Array [
                                "
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bZQynM",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-bZQynM kzlsZH"
                        />
                      </StyledComponent>
                    </styled.div>
                  </Memo()>
                  <Memo()
                    key="1"
                    row={Array []}
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bZQynM",
                              "isStatic": false,
                              "lastClassName": "kzlsZH",
                              "rules": Array [
                                "
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bZQynM",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-bZQynM kzlsZH"
                        />
                      </StyledComponent>
                    </styled.div>
                  </Memo()>
                  <Memo()
                    key="2"
                    row={Array []}
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bZQynM",
                              "isStatic": false,
                              "lastClassName": "kzlsZH",
                              "rules": Array [
                                "
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bZQynM",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-bZQynM kzlsZH"
                        />
                      </StyledComponent>
                    </styled.div>
                  </Memo()>
                  <Memo()
                    key="3"
                    row={Array []}
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bZQynM",
                              "isStatic": false,
                              "lastClassName": "kzlsZH",
                              "rules": Array [
                                "
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bZQynM",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-bZQynM kzlsZH"
                        />
                      </StyledComponent>
                    </styled.div>
                  </Memo()>
                </div>
              </StyledComponent>
            </styled.div>
          </div>
        </StyledComponent>
      </styled.div>
    </Board>
  </Router>
</BrowserRouter>
`;
